ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestinationSync.java

```diff
@@ -680,9 +687,107 @@ public FindResult findEntriesSince(String partitionPath, FindToken findToken, lo
     throw new UnsupportedOperationException("findEntriesSince will not be implemented for AzureCloudDestinationSync");
   }
 
+  /**
+   * Erases blobs from a given list of blobs in cloud
+   * @param blobItemList List of blobs in a container
+   * @param blobContainerClient BlobContainer client
+   * @return The number of blobs erased
+   */
+  protected int eraseBlobs(List<BlobItem> blobItemList, BlobContainerClient blobContainerClient) {
+    int numBlobsPurged = 0;
+    long now = System.currentTimeMillis();
+    long gracePeriod = TimeUnit.DAYS.toMillis(cloudConfig.cloudCompactionGracePeriodDays);
+    for (BlobItem blobItem: blobItemList) {
+      if (shutdownCompaction.get()) {
+        logger.info("Shut down compaction for container {}", blobContainerClient.getBlobContainerName());
+        break;
+      }
+      Map<String, String> metadata = blobItem.getMetadata();
+      boolean eraseBlob = false;
+      String eraseReason = "NONE";
+
+      if (metadata.containsKey(CloudBlobMetadata.FIELD_DELETION_TIME)) {
+        long deletionTime = Long.parseLong(metadata.get(CloudBlobMetadata.FIELD_DELETION_TIME));
+        eraseBlob = (deletionTime + gracePeriod) < now;
+        eraseReason = String.format("%s: (%s + %s) < %s", CloudBlobMetadata.FIELD_DELETION_TIME, deletionTime, gracePeriod, now);
+      } else if (metadata.containsKey(CloudBlobMetadata.FIELD_EXPIRATION_TIME)) {
+        long expirationTime = Long.parseLong(metadata.get(CloudBlobMetadata.FIELD_EXPIRATION_TIME));
+        eraseBlob = (expirationTime + gracePeriod) < now;
+        eraseReason = String.format("%s: (%s + %s) < %s", CloudBlobMetadata.FIELD_EXPIRATION_TIME, expirationTime, gracePeriod, now);
+      }
+
+      if (eraseBlob) {
+        if (cloudConfig.cloudCompactionDryRunEnabled) {
+          logger.trace("[DRY-RUN] Can erase blob {} from Azure blob storage because {}", blobItem.getName(), eraseReason);
+        } else {
+          Timer.Context storageTimer = azureMetrics.blobCompactionLatency.time();
+          logger.trace("Erasing blob {} from Azure blob storage because {}", blobItem.getName(), eraseReason);
+          blobContainerClient.getBlobClient(blobItem.getName()).delete();
```

Comment:
It looks like Azure also has a batch API (which we use in V1) to issue deletes in a batch (and hence save network calls). We could use that too as an optimization. Not urgent though. Just wanted to note.

Reply:
Will consider based on performance measurements. Thank you.